### DISCUSSION TITLE
issue with install fdkaac on RHEL9 (centos stream 9)

### DISCUSSION THREAD

**Discussion by masternet0** [savonet/liquidsoap#2821] [General]
Hello,
I try to install fdkaac by opam for liquidsoap on CentOS stream 9
opam install fdkaac

but there error for this lib 
The following system packages will first need to be installed:
    fdk-aac-devel
    
and i see fdk-aac-devel not available yet on RHEL 9 but replaced with fdk-aac-free-devel
so any fix for this to use fdk-aac-free-devel instead of fdk-aac-devel on RHEL9 ?

any solution ?
Regards

**Comment by toots** [comment]
Hi,

In cases like that you should be able to use the `--assume-depext` option from `opam`:

>        --assume-depexts
>            Skip the installation step for any missing system packages, and
>            attempt to proceed with compilation of the opam packages anyway.
>            If the installation is successful, opam won't prompt again about
>            these system packages. Only meaningful if external dependency
>            handling is enabled.

**Reply by toots** [reply]
Meanwhile, I just pushed a new release with fixed depexts: https://github.com/ocaml/opam-repository/pull/22834

**Reply by masternet0** [reply]
thank you for your response @toots 
but when I install fdkaac with this command 
opam install --assume-depext fdkaac
fdkaac installed but not works properly with liquidsoap

when i added aac+ configuration to liquidsoap for output.icecast not working 
output.icecast(%fdkaac(channels=2,samplerate=44100,bandwidth="auto",bitrate=64),

but mpeg4_aac_lc working 
output.icecast(%fdkaac(channels=2,samplerate=44100,bandwidth="auto",aot="mpeg4_aac_lc",bitrate=128)

how can i fix this 
on centos 8 stream working good both but still on centos 9 stream not working aac+

this is the error logs :

[error.txt](https://github.com/savonet/liquidsoap/files/10344575/error.txt)


Regards

**Reply by toots** [reply]
This looks like an internal issue with your parameters. Not all combination of parameters do work together and also this might change depending on the release. Have you tried the same parameters with the command line encoder?

**Comment by masternet0** [comment]
@toots 
I see in new version fdk-aac 0.3.3 there error in RHEL7 and RHEL8
because this update :
https://github.com/ocaml/opam-repository/pull/22862/commits/eff8353fdad01ce71b19916bbd2167ce2ee688c3

depexts: [
  ["libfdk-aac-dev"] {os-family = "debian" | os-family = "ubuntu"}
  ["fdk-aac-dev"] {os-family = "alpine"}
  ["fdk-aac-free-devel"] {os-distribution="centos" | os-family = "fedora" | os-family = "suse"}
  ["fdk-aac"] {os-distribution = "nixos" | os = "freebsd" | os = "macos" & os-distribution = "homebrew"}
  ["libfdk-aac"] {os-distribution = "arch"}
]

I see you update this lib fdk-aac-free-devel for all RHEL and this is not right 
this lib only replaced on RHEL9 and still old one fdk-aac-devel used by RHEL7 and RHEL8

so please fix this to use fdk-aac-devel on RHEL7 and RHEL8
and fdk-aac-free-devel on RHEL9

Regards

**Reply by toots** [reply]
Hi,

The external dependencies solver cannot really account for OS release versions. The `fdk-aac-free-devel` package seems to be the official one so I went with it.

If you have another package, the process is to install that package and by-pass `opam`'s external dependency checker as explained above.

**Reply by masternet0** [reply]
@toots  
this is not good for auto installation 
because fdk-aac-free-devel not available for  RHEL7 and RHEL8 its only available for RHEL9

this will effect to all software using liquidsoap by opam when start installation, auto install will appear error with new fdk-aac 0.3.3 on RHEL7 and RHEL8
like for example centova, will find centova installation now get error and need fix by centova support to pin fdk-aac 0.3.2 for RHEL7 and RHEL8 and need all softwares team to take action after this update to change installation method for liquidsoap on RHEL7, RHEL8

```
The following actions will be performed:
  ∗ install conf-fdkaac 1     [required by fdkaac]
  ∗ install fdkaac      0.3.3
===== ∗ 2 =====

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[fdkaac.0.3.3] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of conf-fdkaac failed at "/usr/local/centovacast/.opam/opam-init/hooks/sandbox.sh build pkg-config --exists fdk-aac".

#=== ERROR while compiling conf-fdkaac.1 ======================================#
# context     2.0.5 | linux/x86_64 | ocaml-base-compiler.4.08.0 | https://opam.ocaml.org#6698d780
# path        ~/.opam/4.08.0/.opam-switch/build/conf-fdkaac.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build pkg-config --exists fdk-aac
# exit-code   1
# env-file    ~/.opam/log/conf-fdkaac-22166-d2c37b.env
# output-file ~/.opam/log/conf-fdkaac-22166-d2c37b.out



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build conf-fdkaac 1
└─
╶─ No changes have been performed

The packages you requested declare the following system dependencies. Please make sure they are installed before retrying:
    fdk-aac-free-devel

! Error: Failed to build Liquidsoap packages
! Error: Installation failed
Liquidsoap installation failed
! Error: Installer exited with error, aborting
```


so i hope you find way to fix this by your end to define opam use fdk-aac-devel on RHEL7 and RHEL8 and fdk-aac-free-devel on RHEL9
Regards

**Reply by toots** [reply]
I believe that this is should be fairly easy to solve during the automated installation procedure by picking up the package according to the distribution version and then always installing via `opam` without using opam's depext mechanism for it.

While we strive to support as many combination of distributions and architectures, we cannot support all combinations existing or having existed in the past.

Typically here, RHEL 8 was release in 2019, that's already 4 years old and RHEL 7 was released in 2014, almost 10 years ago.

Those distributions get updates but the package set usually stays the same so, if `fdk-aac` was not part of the distribution at the release time, it is still not officially supported as of now, even while there are still security releases for them.

Package name is also an issue for non-free packages which always have renegade, non-official packages with different names and etc.

For these reasons, I believe using the official package name, knowing that in a couple of years all RPM-based distributions officially supporting this package will be using it is a better solution for us long term.

**Comment by gunsar** [comment]
I also did an experiment on Centova Debian, Liquidsoap doesn't show the audio encoder option. it's confusing. can anyone help?
![Screenshot 2023-01-15 083305](https://user-images.githubusercontent.com/56213562/212504716-f8c9e8d3-a031-4218-9ca6-c134d55a6b11.jpg)
